### PR TITLE
Add an implementation of route guide server and client

### PR DESCRIPTION
**Pull request by wonderfly**

### CODE REVIEW

File: examples/route_guide/client/client.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	pb "google.golang.org/grpc/examples/route_guide/proto"
+)
+
+var (
+	tls           = flag.Bool("use_tls", false, "Connection uses TLS if true, else plain TCP")
```
Comment (dsymonds) [inline examples/route_guide/client/client.go]: I meant changing the flag name too. Passing `-tls` on the command-line is just as descriptive as `-use_tls`, except it is less typing.

Either way, the var name here should really match the flag name, so it should be tls/"tls" or useTLS/"use_tls".

File: examples/route_guide/client/client.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+	caFile        = flag.String("tls_ca_file", "testdata/ca.pem", "The file containning the CA root cert file")
```
Comment (dsymonds) [inline examples/route_guide/client/client.go]: same with these other flags: match the var name (modulo converting to camel case) to the flag name.

File: examples/route_guide/client/client.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+	tlsServerName = flag.String("server_host_override", "x.test.youtube.com", "The server name use to verify the hostname returned by TLS handshake")
+)
+
+// printFeature gets the feature for the given point.
+func printFeature(client pb.RouteGuideClient, point *pb.Point) {
+	log.Printf("Getting feature for point (%d, %d)", point.Latitude, point.Longitude)
+	feature, err := client.GetFeature(context.Background(), point)
+	if err != nil {
+		log.Fatalf("%v.GetFeatures(_) = _, %v: ", client, err)
+		return
```
Comment (dsymonds) [inline examples/route_guide/client/client.go]: no need for return after log.Fatalf

File: examples/route_guide/server/server.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+	}
+}
+
+// loadFeatures loads features from a JSON file.
+func (s *routeGuideServer) loadFeatures(filePath string) {
+	file, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		log.Fatal("Failed to load default features: %v", err)
+	}
+	if err := json.Unmarshal(file, &(s.savedFeatures)); err != nil {
```
Comment (dsymonds) [inline examples/route_guide/server/server.go]: unnecessary parens

File: examples/route_guide/server/server.go [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		key := serialize(in.Location)
+		if _, present := s.routeNotes[key]; !present {
+			s.routeNotes[key] = []*pb.RouteNote{in}
+
```
Comment (dsymonds) [inline examples/route_guide/server/server.go]: drop this blank line

File: examples/route_guide/server/server.go [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+func serialize(point *pb.Point) string {
+	return fmt.Sprintf("%d %d", point.Latitude, point.Longitude)
+}
+
+func newServer() *routeGuideServer {
+	s := new(routeGuideServer)
+	s.loadFeatures(*jsonDBFile)
+	s.routeNotes = make(map[string][]*pb.RouteNote, 0)
```
Comment (dsymonds) [inline examples/route_guide/server/server.go]: drop the `, 0`.

**Comment by dsymonds** [comment]
There's my final round of comments.

Also, it looks like the .pb.go file was not checked in. Please do that; we shouldn't force people to have to go install protoc and protoc-gen-go just to build this example.

**Comment by iamqizhao** [comment]
On Thu, Feb 26, 2015 at 3:23 PM, David Symonds <notifications@github.com>
wrote:

> There's my final round of comments.
>
> Also, it looks like the .pb.go file was not checked in. Please do that; we
> shouldn't force people to have to go install protoc and protoc-gen-go just
> to build this example.
>
We kinda like to make this as a real-life example to get readers hands
dirty and be prepared to use gRPC lib.  We deleted the generated code of
this example for all the languages (except  a small number of exceptions --
their codegen are faulty).

File: examples/route_guide/client/client.go (hunk 1 above)
Comment (wonderfly) [inline examples/route_guide/client/client.go]: Agreed. I forgot to update the flag name. Done.

File: examples/route_guide/client/client.go (hunk 2 above)
Comment (wonderfly) [inline examples/route_guide/client/client.go]: Done

File: examples/route_guide/client/client.go (hunk 3 above)
Comment (wonderfly) [inline examples/route_guide/client/client.go]: I was trying not to call log.Println(feature) when there's an error so I returned early. Changed it to if{} else{} to make the flow control more clear.

File: examples/route_guide/server/server.go (hunk 4 above)
Comment (wonderfly) [inline examples/route_guide/server/server.go]: Done

File: examples/route_guide/server/server.go (hunk 5 above)
Comment (wonderfly) [inline examples/route_guide/server/server.go]: Done

File: examples/route_guide/server/server.go (hunk 6 above)
Comment (wonderfly) [inline examples/route_guide/server/server.go]: Done

**Comment by wonderfly** [comment]
Addressing comments. Looks like the generated code got checked in after all as it's required for Travis tests.

**Comment by wonderfly** [comment]
GitHub is not picking up my latest commits because my fork stopped working as the upstream went public and I recreated it. Should I start a new PR instead?

**Comment by iamqizhao** [comment]
Yup, refork and create a new PR please.

**Pull request merged**